### PR TITLE
TreeNodeInfoInit, LSRA, and more over LIR

### PR DESCRIPTION
**Pull request by pgavlin**
With these changes I am able to successfully compile and run a simple Hello, World! application. The result is not diff-free w.r.t. an unmodified build from master, but it does at least run without crashing.

### CODE REVIEW

**Comment by pgavlin** [comment]
@BruceForstall PTAL

File: src/jit/lir.h [hunk 1]
```diff
@@ -25,9 +25,17 @@ class LIR final
         enum : unsigned char
         {
             None = 0x00,
+
             Mark = 0x01, // An aribtrary "mark" bit that can be used in place of
                          // a more expensive data structure when processing a set
                          // of LIR nodes. See for example `LIR::GetTreeRange`.
+
+            IsUnusedValue = 0x02, // Set on a node if it produces a value that is noy
```
Comment (BruceForstall) [inline src/jit/lir.h]: > noy [](start = 82, length = 3)

typo: noy

File: src/jit/lir.h [hunk 2]
```diff
@@ -25,9 +25,17 @@ class LIR final
...
+                                  // subsequently used. Should never be set on nodes
+                                  // that return `false` for `GenTree::IsValue`. Note
+                                  // that this bit should not be assumed to be valid
+                                  // at all points during compilation: it is currently
+                                  // only computed during taget-dependent lowering.
```
Comment (BruceForstall) [inline src/jit/lir.h]: > taget [](start = 58, length = 5)

typo: taget

File: src/jit/flowgraph.cpp [hunk 3]
```diff
@@ -8768,10 +8768,16 @@ BasicBlock* Compiler::fgSplitBlockAfterNode(BasicBlock* curr, GenTree* node)
 
     if (node != nullptr)
     {
-        newBlock->bbTreeList = node->gtNext;
-        newBlock->bbLastNode = curr->bbLastNode;
+        LIR::Range currBBRange = LIR::AsRange(curr);
+        LIR::Range newBlockRange = LIR::AsRange(newBlock);
+
```
Comment (BruceForstall) [inline src/jit/flowgraph.cpp:8772]: could sink this into the "if"

File: src/jit/flowgraph.cpp [hunk 4]
```diff
@@ -17975,12 +17987,19 @@ void                Compiler::fgSetTreeSeqHelper(GenTreePtr tree)
...
 }
 
 void
-Compiler::fgSetTreeSeqFinish(GenTreePtr tree)
+Compiler::fgSetTreeSeqFinish(GenTreePtr tree, bool isLIR)
 {
+    // If we are sequencing a node that does not appear in LIR,
+    // do not add it to the list.
+    if (isLIR && ((tree->OperGet() == GT_LIST) || tree->OperGet() == GT_ARGPLACE))
+    {
```
Comment (BruceForstall) [inline src/jit/flowgraph.cpp:17998]: Shouldn't this just be a check on GTK_NOTLIR? though it looks like you don't have GT_LIST and GT_ARGPLACE in that set for some reason.

Comment (pgavlin) [inline src/jit/flowgraph.cpp:17998]: > Shouldn't this just be a check on GTK_NOTLIR? 

Not quite. If we're sequencing a tree with `isLIR` set, we really expect all of the nodes in the tree to be LIR (i.e. `GenTree::IsLIR()` should return true for each node). Indeed, if we insert a non-LIR node into an LIR range, `CheckLIR()` will assert. With the current design, `GT_LIST` and `GT_ARGPLACE` are still allowable in LIR trees, just not in LIR blocks.

> though it looks like you don't have GT_LIST and GT_ARGPLACE in that set for some reason.

See above.

File: src/jit/lower.cpp [hunk 5]
```diff
@@ -2321,10 +2323,14 @@ void Lowering::InsertPInvokeMethodProlog()
     store->gtOp.gtOp1 = call;
     store->gtFlags |= GTF_VAR_DEF;
 
-    GenTree* insertionPoint = m_blockRange.FirstNonPhiOrCatchArgNode();
+    GenTree* insertionPoint = firstBlockRange.FirstNonPhiOrCatchArgNode();
+    if (insertionPoint == nullptr)
+    {
+        insertionPoint = firstBlockRange.EndExclusive();
```
Comment (BruceForstall) [inline src/jit/lower.cpp:2329]: I'm not liking this name very much anymore. Seems like we should have First == first node, Last == last node, Begin and End should only be on iterators.

File: src/jit/flowgraph.cpp (hunk 4 above)
Comment (pgavlin) [inline src/jit/flowgraph.cpp:17998]: I should also point out that `GTK_NOTLIR` isn't the sole factor is whether or not a node is well-formed LIR: `GenTree::IsLIR()` contains a few other checks. It does in fact enforce that if a `GT_LIST` or `GT_ARGPLACE` node is encountered in an LIR tree that the node is not linked into the containing block.

File: src/jit/lower.cpp (hunk 5 above)
Comment (pgavlin) [inline src/jit/lower.cpp:2329]: I agree. I've really been finding the naming confusing myself. Let's do that as a follow-up change.

**Comment by pgavlin** [comment]
Regarding diffs: there are some assembly diffs between the baseline backend and the LIR backend when compiling my test program (which is just `class A { static void Main() { System.Console.WriteLine("Hello, World!"); } }`). All of these diffs are caused by differences in liveness when handling  trees marked `GTF_VAR_USEDEF` or attempting dead store removal. In the case of the former, the LIR backend is more conservative and overestimates liveness; in the case of the latter, the LIR backend is a bit more aggressive due to analysis improvements made possible by removing embedded statements.

**Comment by BruceForstall** [comment]
I still think we need to get to zero diffs before checking in, even if that means adding pessimizing code temporarily.

---

In reply to: [237280371](https://github.com/pgavlin/coreclr/pull/9#issuecomment-237280371) [](ancestors = 237280371)

**Comment by BruceForstall** [comment]
LGTM

File: src/jit/lir.h (hunk 1 above)
Comment (pgavlin) [inline src/jit/lir.h]: Fixed.

File: src/jit/lir.h (hunk 2 above)
Comment (pgavlin) [inline src/jit/lir.h]: Fixed.

File: src/jit/flowgraph.cpp (hunk 3 above)
Comment (pgavlin) [inline src/jit/flowgraph.cpp:8772]: Done.

**Pull request merged**